src/graaflib/algorithm/shortest_path.tpp

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include <algorithm>
+#include <queue>
+#include <unordered_set>
+#include <unordered_map>
+
+namespace graaf::algorithm {
+
+namespace detail {
+
+template <typename V, typename E, graph_spec S>
+GraphPath<int> get_unweighted_shortest_path(const graph<V, E, S>& graph,
+                                            vertex_id_t start_vertex,
+                                            vertex_id_t end_vertex) {
+  std::unordered_set<vertex_id_t> seen_vertices{};
+  std::unordered_map<vertex_id_t, vertex_id_t> prev_vertex{};
+  std::queue<vertex_id_t> to_explore{};
+
+  to_explore.push(start_vertex);
+  seen_vertices.insert(start_vertex);
+
+  while (!to_explore.empty()) {
+    auto current{to_explore.front()};
+    to_explore.pop();
+
+    if (current == end_vertex) {
+      break;
+    }
+
+    for (const auto& neighbor : graph.get_neighbors(current)) {
+      if (!seen_vertices.contains(neighbor)) {
+        seen_vertices.insert(neighbor);
+        prev_vertex[neighbor] = current;
+        to_explore.push(neighbor);
+      }
+    }
+  }
+
+  if (seen_vertices.contains(end_vertex)) {
+    GraphPath<int> path;
+    auto current{end_vertex};
+
+    while (current != start_vertex) {
+      path.vertices.push_back(current);
+      current = prev_vertex[current];
+    }
+
+    path.vertices.push_back(start_vertex);
+    reverse(path.vertices.begin(), path.vertices.end());
+    path.total_weight = path.vertices.size();
+
+    return path;
+
+  } else {
+    return GraphPath<int>{{}, -1};
+  }
+}
+
+}  // namespace detail
+
+template <edge_strategy EDGE_STRATEGY, typename V, typename E, graph_spec S>
+GraphPath<E> get_shortest_path(const graph<V, E, S>& graph,
+                               vertex_id_t start_vertex,
+                               vertex_id_t end_vertex) {
+  using enum edge_strategy;
+  if constexpr (EDGE_STRATEGY == UNWEIGHTED) {
+    return detail::get_unweighted_shortest_path(graph, start_vertex,
+                                                end_vertex);
+  }
+
+  if constexpr (EDGE_STRATEGY == WEIGHTED) {
+    // TODO: Implement A* or Dijkstra
+    throw std::logic_error(
+        "Shortest path for weighted graphs not yet implemented.");
+  }
```

Comment:
I think the weighted implementation is going to be a bit hard currently. A `graph`'s `EDGE_T` is user provided and can be an arbitrary class. Hence there is no clear weight of an edge defined for any `EDGE_T` other than a numeric primitive. 

One solution would be to provide an abstract `weighted_edge` which users could inherit from when defining their edge type. This class then has to provide a method such as `get_weight()`. If you want I could open a PR for this, wdyt?

Reply:
Fully agree. With the current interface, it will only work if `EDGE_T` is just a numeric type. I think this is fair for a lot of use cases already, but obviously introduces limits.
There should be a dedicated PR for that since there are some caveats in your proposal. Example: What if there is no single or default edge weight in the `EDGE_T` class, but multiple?  For the distance case, this might be weights `time_by_train` and `time_by_car`.

Reply:
+ as you already wrote elsewhere, this also makes a case for merging this PR without the weighted logic already to not drag this PR too long